tsl/src/compression/compression.c

```diff
@@ -1898,6 +1901,88 @@ build_scankeys(int32 hypertable_id, RowDecompressor decompressor, Bitmapset *key
 					key_index++;
 				}
 			}
+			if (COMPRESSIONCOL_IS_ORDER_BY(fd))
+			{
+				bool isnull;
+				Datum value = slot_getattr(slot, attno, &isnull);
+
+				/* Cannot optimize orderby columns with NULL values since those
+				 * are not visible in metadata
+				 */
+				if (isnull)
+					continue;
+
+				char *segment_min_col_name = compression_column_segment_min_name(fd);
+				char *segment_max_col_name = compression_column_segment_max_name(fd);
+				segment_min_attr_number =
+					get_attnum(decompressor.in_rel->rd_id, segment_min_col_name);
+				Assert(segment_min_attr_number != InvalidAttrNumber);
+				segment_max_attr_number =
+					get_attnum(decompressor.in_rel->rd_id, segment_max_col_name);
+				Assert(segment_max_attr_number != InvalidAttrNumber);
+
+				/* This should never happen but if it does happen, we can't generate a scan key for
+				 * the orderby column so just skip it */
+				if (segment_min_attr_number == InvalidAttrNumber ||
+					segment_max_attr_number == InvalidAttrNumber)
+					continue;
+
+				/* Orderby column type should match in compressed metadata columns and uncompressed
+				 * chunk attribute */
+				Assert(decompressor.out_desc->attrs[AttrNumberGetAttrOffset(attno)].atttypid ==
+					   decompressor.in_desc->attrs[AttrNumberGetAttrOffset(segment_min_attr_number)]
+						   .atttypid);
+				Assert(decompressor.out_desc->attrs[AttrNumberGetAttrOffset(attno)].atttypid ==
+					   decompressor.in_desc->attrs[AttrNumberGetAttrOffset(segment_max_attr_number)]
+						   .atttypid);
+
+				Oid atttypid =
+					decompressor.out_desc->attrs[AttrNumberGetAttrOffset(attno)].atttypid;
+				TypeCacheEntry *tce = lookup_type_cache(atttypid, TYPECACHE_BTREE_OPFAMILY);
+				if (!OidIsValid(tce->btree_opf))
+					elog(ERROR, "no btree opfamily for type \"%s\"", format_type_be(atttypid));
+
+				le_opr = get_opfamily_member(tce->btree_opf,
+											 atttypid,
+											 atttypid,
+											 BTLessEqualStrategyNumber);
+				ge_opr = get_opfamily_member(tce->btree_opf,
+											 atttypid,
+											 atttypid,
+											 BTGreaterEqualStrategyNumber);
+				Assert(OidIsValid(le_opr));
+				Assert(OidIsValid(ge_opr));
```

Comment:
there's a lot of duplication here...

can't you reorganize this logic to reduce that?

one idea I could think of is something like:
```
if( init_inequality_scankey(&scankeys[key_index+0], BTLessEqualStrategyNumber, ...) &&
    init_inequality_scankey(&scankeys[key_index+1],BTGreaterEqualStrategyNumber,... ) {

  key_index+=2;
}
```